packages/logger/src/logger/PersistentTransport.ts

```diff
@@ -0,0 +1,111 @@
+
+    const botIdentifier = process.env.BOT_IDENTIFIER || noBotId;
+    this.logListKey = `persistentLogQueue:${botIdentifier}:${derivedTransport}`;
+  }
+
+  // Getter for checking if the transport is flushed.
+  get isFlushed(): boolean {
+    return !this.isQueueBeingExecuted;
+  }
+
+  // Note: info must be any because that's what the base class uses.
+  async log(info: any, callback: (error?: unknown) => void): Promise<void> {
+    try {
+      // We only support persisting dictionary object logs.
+      if (!isDictionary(info)) throw new Error("Unsupported info type!");
+
+      await this.connectRedis();
+
+      await this.redis.rPush(this.logListKey, JSON.stringify(info));
+    } catch (error) {
+      return callback(new TransportError(this.derivedTransport, error, info));
+    }
+    // Initiate log que processing. We don't await it as this should run in background and it is controlled externally
+    // via pauseProcessing method.
+    this.processLogQueue();
+
+    callback();
+  }
+
+  // Signal to pause processing messages from persistent storage.
+  pauseProcessing() {
+    this.canProcess = false;
+  }
+
+  // Logs queue element when processing persistent storage. Implementation is specific to derived class.
+  abstract logQueueElement(info: Record<string, unknown>): Promise<void>;
+
+  // Processes log queue from persistent storage.
+  async processLogQueue(): Promise<void> {
+    // Avoid concurrent log queue processing and lock it.
+    if (this.isQueueBeingExecuted) return;
+    this.isQueueBeingExecuted = true;
+
+    // We will need access to info in catch block for more verbose error handling.
+    let info: Record<string, unknown> | null;
+
+    // Process queue unless received signal to pause before termination.
+    while (this.canProcess) {
+      info = null; // Reset info at the beginning of each iteration.
+
+      try {
+        await this.connectRedis();
+
+        const oldestLogString = await this.redis.lPop(this.logListKey);
+        if (oldestLogString === null) break; // We have processed all logs from persistent storage queue.
+
+        info = JSON.parse(oldestLogString);
+        if (!isDictionary(info)) throw new Error("Unsupported info type!");
+
+        // Log the message in the derived transport implementation.
+        await this.logQueueElement(info);
+      } catch (error) {
+        if (info === null) {
+          // We cannot emit TransportError as we don't have access to original info object yet.
+          this.emit("error", error);
+        } else this.emit("error", new TransportError(this.derivedTransport, error, info));
+        break;
+      }
+
+      // Wait before processing the next message if the implementation requires any rate limiting.
+      await delay(this.rateLimit);
+    }
+
+    // Unlock the queue execution.
+    this.isQueueBeingExecuted = false;
+  }
+
+  // Connect to redis when not ready.
+  private async connectRedis(): Promise<void> {
```

Comment:
another option here is to initialize redis at a higher level where you can await it, and pass the connected client in as a dependency. its a little cleaner since then you avoid a scenario where you may forget to run this at the top of a call.  not suggesting a refactor but just wanted to point it out

Reply:
Its bit tricky as app layer is not aware of what transports get attached to its logger as that gets determined in logger constructor based on passed environment variables.